### DISCUSSION TITLE
Varible types change when creating table

### DISCUSSION THREAD

**Discussion by ironsoul** [sqlalchemy/sqlalchemy#7622] [Usage Questions]
When I create tables its not creating an exact copy. The variable types are different, for example the int8 is changed to bigint which is smaller than int8. Therefor when I do an insert i get the index out of range error. Is there a way to cater for this?

The code I am using to create the tables are as follows:

```
for current_table in tbl_to_load:
    conn_source = create_engine(db_connection_string, echo=False, echo_pool=False)
    conn_destination = create_engine(db_connection_string, echo=False, echo_pool=False)


    meta_source = MetaData(schema = schema_source)
    table_1 = Table(current_table,
                    meta_source ,
                    autoload_with=conn_source 
                   )
    
        
    drop_table = text('DROP TABLE IF EXISTS {}."{}"'.format(schema_destination, current_table))
    conn_destination.execute(drop_table)
    
    
    conn2 = conn_destination.connect().execution_options(schema_translate_map={'public': "schema_name"})
    table_1.create(bind=conn2, checkfirst=True)


    meta_destination = MetaData(schema = schema_destination)
    table_2 = Table(current_table,
                    meta_destination ,
                    autoload_with=conn_destination 
                    )
```

Then I go onto loading the data.

**Comment by CaselIT** [comment]
Hi,

Which database is it?

**Reply by ironsoul** [reply]
Both the source database and destination database are postgresql.

**Reply by CaselIT** [reply]
> the int8 is changed to bigint which is smaller than int8. Therefor when I do an insert i get the index out of range error.

that does not seems right. from https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-INT bigint is an alias of int8 in postgresql. could you post the output of psql `\d` of the relevant columns of the two tables?

**Reply by zzzeek** [reply]
if you need to adjust types quickly, use the [column_reflect](https://docs.sqlalchemy.org/en/14/core/events.html#sqlalchemy.events.DDLEvents.column_reflect) event to intercept and correct these for now.

**Reply by ironsoul** [reply]
After some further investigation its not the int8/bigint that was the issue. It was on the insert, I wasn't handling the null values correctly. It seems like I get the error index out of range when trying to insert null values. This also happened with a timestamp field where the error was: invalid input syntax for type timestamp: "NaT". I was able to fix it by changing the way I do the insert.

**Comment by ironsoul** [comment]
I have a follow-up question here. I want to use the same code as above (to create tables in the destination db from a source db) but the source ddl is now coming from a MySQL db. MySQL uses Tinyint and datetime which is not used in postgresql. Is there a way to change the Tinyint to int on the fly, or am I going to need to do something more elaborate?

Error I get: Compiler <sqlalchemy.dialects.postgresql.base.PGTypeCompiler object at 0x00000...> can't render element of type TINYINT'

I am using pymysql
Source db MySQL
Destination db Postgresql

**Reply by CaselIT** [reply]
You can use the [column_reflect](https://docs.sqlalchemy.org/en/14/core/events.html#sqlalchemy.events.DDLEvents.column_reflect) event as suggested by mike here https://github.com/sqlalchemy/sqlalchemy/discussions/7622#discussioncomment-2050870

**Reply by ironsoul** [reply]
I have been trying to follow this example to reflect my columns but still getting errors [Reflecting with Database-Agnostic Types](https://docs.sqlalchemy.org/en/14/core/reflection.html#reflecting-with-database-agnostic-types) 

My code:
```
from sqlalchemy import MetaData, Table, create_engine
mysql_engine = create_engine("mysql+pymysql://user_name:password@IP:Port", echo=True)
metadata_obj = MetaData()
my_mysql_table = Table("table_name", metadata_obj, schema='schema_name', autoload_with=mysql_engine)

from sqlalchemy.schema import CreateTable
print(CreateTable(my_mysql_table).compile(mysql_engine))

from sqlalchemy import event
metadata_obj = MetaData()

@event.listens_for(metadata_obj, "column_reflect")
def genericize_datatypes(inspector, tablename, column_dict):
    column_dict["type"] = column_dict["type"].as_generic()

		##fails on the listener

my_generic_table = Table("table_name", metadata_obj, schema='schema_name', autoload_with=mysql_engine)

pg_engine = create_engine("postgresql://user_name:password@IP:Port", echo=True)
my_generic_table.create(pg_engine)
```

That MySQL tables variables:
- Varchar
- Int
- Datatime
- Tinyint and Tinyint(1)

Error occurs when listener part is run. Error:
NotImplementedError: Default TypeEngine.as_generic() heuristic method was unsuccessful for sqlalchemy.dialects.mysql.types.BIT. A custom as_generic() method must be implemented for this type class.

I am not sure what to do to fix this. I did see  #6345 But couldn't find a solution in there. Not sure if I am just missing something.

**Reply by zzzeek** [reply]
for that one type you need to pick a datatype that you want to transfer over, such as Boolean :

```py
@event.listens_for(metadata_obj, "column_reflect")
def genericize_datatypes(inspector, tablename, column_dict):
    try:
        column_dict["type"] = column_dict["type"].as_generic()
    except NotImplementedError:
        if isinstance(column_dict["type"],mysql.BIT):
            column_dict["type"] = Boolean()
        else:
            raise
```

**Reply by CaselIT** [reply]
Maybe we could add a note on that part of the docs.

I guess mapping it to Boolean does bot work?

**Reply by ironsoul** [reply]
Thank you for that, I got it working. I mapped mine to an integer.

**Reply by zzzeek** [reply]
MySQL TINYINT is the traditional boolean datatype, BIT might be used that way here but that's not the canonical practice.

**Comment by ironsoul** [comment]
I'm back again with another question. Is it alright for me to keep this open and keep asking questions related to tables and columns in this discussion or should i close this one and open one per question? 

So the question:
Following on the stuff I was doing above (moving data from MySQL to Postgresql) I ran into the following error. 
ValueError: A string literal cannot contain NUL (0x00) characters.

I have tried a few things and its not working, once again any advise will be appreciated. I create the table so that each column accepts nulls, but still getting the same error. I also thought it might be my insert, so I tried converting it to a dataframe and then use the to_sql function, but also got the same error.

For reference, my insert code:

```Python
    while 'batch not empty':  
        batch = proxy.fetchmany(10000)  
        list1 = []

        if not batch:
            break

        for row in batch:
            d = dict(row._mapping.items())
            list1.append(d)    
        
        conn_sl.execute(table_2.insert(), list1)  
```

**Reply by CaselIT** [reply]
As you prefer for the discussion, I don't think they can be closed. For unrelated things maybe it's better a new discussion, but this seems related enough.

Regarding the question, are you sure you are not moving binary data? What postgresql is complaining about is that there are null chars (0x00) in the data, and that's not valid for the destination column. That's a different thing that sql null

**Comment by ironsoul** [comment]
After further investigation, I found that some of the records has ![image](https://user-images.githubusercontent.com/97736952/154225879-50f5a558-9c9d-412e-abe8-3070f947b69f.png) in their varchar columns. The nul appears in the middle of a varchar not on its own.

Is there a way to cater for these nul's? To replace or ignore them?

**Reply by CaselIT** [reply]
> Is there a way to cater for these nul's? To replace or ignore them?

this really depends on your use case. from here https://www.postgresql.org/docs/current/datatype-character.html nul chars cannot be stored in text columns in postgresql. I think you will need to use binary columns if you want to keep them

**Reply by ironsoul** [reply]
Thank you, I will read up on that. So the currently I'm just moving data from MySQL to Postgresql (lift and shift). I am happy to exclude, remove or convert the binary entries.
